lib/luacrypto_skcipher.c

```diff
@@ -0,0 +1,248 @@
+* @function crypto_skcipher:encrypt
+* @tparam string iv The Initialization Vector. Its length must match `ivsize()`.
+* @tparam string plaintext The data to encrypt.
+* @treturn string The ciphertext.
+* @raise Error on encryption failure, incorrect IV length, or allocation issues.
+*/
+LUACRYPTO_SKCIPHER_CRYPT_FN(encrypt);
+
+/***
+* Decrypts ciphertext using the SKCIPHER transform.
+* The IV must match the one used during encryption.
+* Ciphertext length should be appropriate for the cipher mode.
+* @function crypto_skcipher:decrypt
+* @tparam string iv The Initialization Vector. Its length must match `ivsize()`.
+* @tparam string ciphertext The data to decrypt.
+* @treturn string The plaintext.
+* @raise Error on decryption failure, incorrect IV length, or allocation issues.
+*/
+LUACRYPTO_SKCIPHER_CRYPT_FN(decrypt);
+
+/***
+* Lua C methods for the SKCIPHER TFM object.
+* Includes cryptographic operations and Lunatik metamethods.
+* The `__close` method is important for explicit resource cleanup.
+* @see crypto_skcipher
+* @see lunatik_closeobject
+*/
+
+static const luaL_Reg luacrypto_skcipher_mt[] = {
+	{"setkey", luacrypto_skcipher_setkey},
+	{"ivsize", luacrypto_skcipher_ivsize},
+	{"blocksize", luacrypto_skcipher_blocksize},
+	{"encrypt", luacrypto_skcipher_encrypt},
+	{"decrypt", luacrypto_skcipher_decrypt},
+	{"__gc", lunatik_deleteobject},
+	{"__close", lunatik_closeobject},
+	{"__index", lunatik_monitorobject},
+	{NULL, NULL}
+};
+
+/***
+* Lunatik class definition for SKCIPHER TFM objects.
+* This structure binds the C implementation (luacrypto_skcipher_t, methods, release function)
+* to the Lua object system managed by Lunatik.
+*/
+
+static const lunatik_class_t luacrypto_skcipher_class = {
+	.name = "crypto_skcipher",
+	.methods = luacrypto_skcipher_mt,
+	.release = luacrypto_skcipher_release,
+	.sleep = true,
+};
+
+/***
+* Creates a new SKCIPHER transform (TFM) object.
+* This is the constructor function for the `crypto_skcipher` module.
+* @function crypto_skcipher.new
+* @tparam string algname The name of the skcipher algorithm (e.g., "cbc(aes)", "ctr(aes)").
+* @treturn crypto_skcipher The new SKCIPHER TFM object.
+* @raise Error if the TFM object or kernel request cannot be allocated/initialized.
+* @usage
+*   local skcipher_mod = require("crypto_skcipher")
+*   local cipher = skcipher_mod.new("cbc(aes)")
+*/
+static int luacrypto_skcipher_new(lua_State *L) {
+	const char *algname = luaL_checkstring(L, 1);
+	luacrypto_skcipher_t *tfm_ud;
+	gfp_t gfp = lunatik_gfp(lunatik_toruntime(L));
+	lunatik_object_t *object = lunatik_newobject(L, &luacrypto_skcipher_class, sizeof(luacrypto_skcipher_t));
+	tfm_ud = (luacrypto_skcipher_t *)object->private;
+
+	tfm_ud->tfm = crypto_alloc_skcipher(algname, 0, 0);
+	if (IS_ERR(tfm_ud->tfm)) {
+		long err = PTR_ERR(tfm_ud->tfm);
+		return luaL_error(L, "failed to allocate SKCIPHER transform for %s (err %ld)", algname, err);
```

Comment:
we are not using this pattern `return luaL_error()` in lunatik because it raises warnings in kbuild if I'm not mistaken.. can you check if it's the case.. I'm not against using it in lunatik if it doesn't raise unnecessary warnings, but I would apply it globally not in a single module.. one of the things I think we should stick to is the uniformity of the code base.. I'm not against on changes as long as we keep the framework concise =)